src/lib/otp_lib/agent.mli

```diff
@@ -0,0 +1,17 @@
+module type Read_only_intf = sig
+  type 'a t
+
+  val get : 'a t -> 'a
+
+  val on_update : 'a t -> f:('a -> unit) -> unit
+end
+
+module Read_only : Read_only_intf
+
+include Read_only_intf
+
+val create : f:('a -> 'b) -> 'a -> 'b t
+
+val update : 'a t -> 'a -> unit
+
+val read_only : 'a t -> 'a Read_only.t
```

Comment:
Using a GADT would be preferable to opaque module type inequality, but this works. However, for reference, you can define something like this:

```ocaml
type read_write
type read_only
type _ flag = Read_write : read_write flag | Read_only : read_only flag
type t_
type ('flag, 't) t = t_ constraint 'flag = _ flag

val create : f:('a -> 'b) -> 'a -> (read_write flag, 'b) t
val read : (_ flag, 'a) t -> 'a
val write : (read_write flag, 'a) t -> 'a -> unit
```

Reply:
yeah that's better